src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs

```diff
@@ -818,10 +820,16 @@ private NamedTypeSymbol ApplyGenericArguments(NamedTypeSymbol symbol, Type[] typ
             }
 
             // Lookup in references
+            bool onlyConsiderCorLib = false;
```

Comment:
>bool onlyConsiderCorLib = false; [](start = 12, length = 32)

It feels like the implementation could be simpler and, perhaps, faster for scenarios when the type is indeed defined in core library. I would expect it roughly structured this way:
If we give priority to core library among other references and there are references (i.e. we are definitely not inside a core library, otherwise we already performed the lookup above), look in the ```CorLibrary```. If found, that is our result. If not found, continue with lookup in all references, skipping the ```CorLibrary``` as an optimization.

It is quite possible I am missing some some detail and the suggested structure wouldn't work for some specific scenario. That is why, I think, we need to state the lookup rules in words here, clearly and precisely explaining  the lookup rules and what gets priority and when. Also, we should clearly explain how ```ignoreCorLibraryDuplicatedTypes```is supposed to affect the lookup, because it appears to be in conflict with a general rule.

Reply:
In the process, it might be worth checking if ```CorLibrary``` is missing.

---
In reply to: [585718223](https://github.com/dotnet/roslyn/pull/51571#discussion_r585718223) [](ancestors = 585718223)